File: csharp/Praxis/src/Core/PluresDB/SchemaRegistry.cs
// <copyright file="SchemaRegistry.cs" company="Plures">
// Copyright (c) Plures. All rights reserved.
// Licensed under the MIT License.
// </copyright>

namespace Praxis.Core.PluresDB;

/// <summary>
/// A simplified Praxis schema definition.
/// </summary>
/// <param name="Name">The schema name.</param>
/// <param name="Version">The schema version.</param>
/// <param name="Description">Optional description.</param>
public sealed record PraxisSchema(
    string Name,
    string Version,
    string? Description = null);

/// <summary>
/// Stored schema entry with metadata.
/// </summary>
/// <param name="Schema">The schema definition.</param>
/// <param name="RegisteredAt">When the schema was registered.</param>
/// <param name="Version">Schema version.</param>
public sealed record StoredSchema(
    PraxisSchema Schema,
    long RegisteredAt,
    string Version);

/// <summary>
/// PraxisSchemaRegistry manages schema definitions in PluresDB.
/// Schemas are stored under /_praxis/schemas/schemaName.
/// </summary>
public sealed class PraxisSchemaRegistry
{
    private readonly IPraxisDB _db;

    /// <summary>
    /// Initializes a new instance of the <see cref="PraxisSchemaRegistry"/> class.
    /// </summary>
    /// <param name="db">The PraxisDB instance.</param>
    public PraxisSchemaRegistry(IPraxisDB db)
    {
        _db = db;
    }

    /// <summary>
    /// Gets the path for a schema in PluresDB.
    /// </summary>
    /// <param name="schemaName">The schema name.</param>
    /// <returns>The schema path.</returns>
    public static string GetSchemaPath(string schemaName) => $"{PraxisPaths.Schemas}/{schemaName}";

    /// <summary>
    /// Registers a schema in PluresDB.
    /// </summary>
    /// <param name="schema">The schema to register.</param>
    /// <returns>A task representing the async operation.</returns>
    public async Task RegisterAsync(PraxisSchema schema)
    {
        var path = GetSchemaPath(schema.Name);

        var storedSchema = new StoredSchema(
            Schema: schema,
            RegisteredAt: DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            Version: schema.Version);

        await _db.SetAsync(path, storedSchema);
    }

    /// <summary>
    /// Gets a schema by name.
    /// </summary>
    /// <param name="schemaName">The schema name.</param>
    /// <returns>The stored schema or null if not found.</returns>
    public async Task<StoredSchema?> GetAsync(string schemaName)
    {
        var path = GetSchemaPath(schemaName);
        return await _db.GetAsync<StoredSchema>(path);
    }

    /// <summary>
    /// Checks if a schema is registered.
    /// </summary>
    /// <param name="schemaName">The schema name.</param>
    /// <returns>True if the schema exists.</returns>
    public async Task<bool> ExistsAsync(string schemaName)
    {
        var stored = await GetAsync(schemaName);
        return stored != null;
    }

    /// <summary>
    /// Updates a schema (replaces existing).
    /// </summary>
    /// <param name="schema">The updated schema.</param>
    /// <returns>A task representing the async operation.</returns>
    public Task UpdateAsync(PraxisSchema schema) => RegisterAsync(schema);

    /// <summary>
    /// Lists all registered schema names.
    /// Implementation note: This method uses an index stored at /_praxis/schemas/_index.
    /// When using InMemoryPraxisDB, schemas must be registered using RegisterWithIndexAsync()
    /// for them to appear in this listing.
    /// </summary>
    /// <returns>Array of registered schema names.</returns>
    public async Task<IReadOnlyList<string>> ListAsync()
    {
        var indexPath = $"{PraxisPaths.Schemas}/_index";
        var index = await _db.GetAsync<List<string>>(indexPath);
        return index ?? [];
    }

    /// <summary>
    /// Registers a schema and updates the index.
    /// </summary>
    /// <param name="schema">The schema to register.</param>
    /// <returns>A task representing the async operation.</returns>
    public async Task RegisterWithIndexAsync(PraxisSchema schema)
    {
        // Register the schema
        await RegisterAsync(schema);

        // Update the index
        var indexPath = $"{PraxisPaths.Schemas}/_index";
        var existingIndex = await _db.GetAsync<List<string>>(indexPath) ?? [];

        if (!existingIndex.Contains(schema.Name))
        {
            existingIndex.Add(schema.Name);
            await _db.SetAsync(indexPath, existingIndex);
        }
    }
}

/// <summary>
/// Helper methods for schema registration.
/// </summary>
public static class SchemaRegistryHelper
{
    /// <summary>
    /// Registers a schema in PluresDB.
    /// Convenience function for one-off schema registration.
    /// </summary>
    /// <param name="db">The PraxisDB instance.</param>
    /// <param name="schema">The schema to register.</param>
    /// <returns>A task representing the async operation.</returns>
    /// <example>
    /// <code>
    /// var db = PraxisDBFactory.CreateInMemoryDB();
    /// await SchemaRegistryHelper.RegisterSchemaAsync(db, new PraxisSchema(
    ///     Name: "MyApp",
    ///     Version: "1.0.0",
    ///     Description: "My application schema"));
    /// </code>
    /// </example>
    public static async Task RegisterSchemaAsync(IPraxisDB db, PraxisSchema schema)
    {
        var registry = new PraxisSchemaRegistry(db);
        await registry.RegisterWithIndexAsync(schema);
    }

    /// <summary>
    /// Creates a PraxisSchemaRegistry instance.
    /// </summary>
    /// <param name="db">The PraxisDB instance.</param>
    /// <returns>PraxisSchemaRegistry instance.</returns>
    public static PraxisSchemaRegistry CreateSchemaRegistry(IPraxisDB db) => new(db);
}

File: src/core/pluresdb/schema-registry.ts
/**
 * Praxis Schema Registry for PluresDB
 *
 * Registers Praxis schema definitions in PluresDB for type-safe storage
 * and cross-agent schema discovery.
 */

import type { PraxisDB } from './adapter.js';
import type { PraxisSchema } from '../schema/types.js';
import { PRAXIS_PATHS } from './store.js';

/**
 * Get the path for a schema in PluresDB
 * @param schemaName The schema name
 */
export function getSchemaPath(schemaName: string): string {
  return `${PRAXIS_PATHS.SCHEMAS}/${schemaName}`;
}

/**
 * Stored schema entry with metadata
 */
export interface StoredSchema {
  /** The schema definition */
  schema: PraxisSchema;
  /** When the schema was registered */
  registeredAt: number;
  /** Schema version */
  version: string;
}

/**
 * PraxisSchemaRegistry
 *
 * Manages schema definitions in PluresDB.
 * Schemas are stored under `/_praxis/schemas/<schemaName>`
 */
export class PraxisSchemaRegistry {
  private db: PraxisDB;

  constructor(db: PraxisDB) {
    this.db = db;
  }

  /**
   * Register a schema in PluresDB
   *
   * @param schema The schema to register
   */
  async register(schema: PraxisSchema): Promise<void> {
    const path = getSchemaPath(schema.name);

    const storedSchema: StoredSchema = {
      schema,
      registeredAt: Date.now(),
      version: schema.version,
    };

    await this.db.set(path, storedSchema);
  }

  /**
   * Get a schema by name
   *
   * @param schemaName The schema name
   * @returns The stored schema or undefined if not found
   */
  async get(schemaName: string): Promise<StoredSchema | undefined> {
    const path = getSchemaPath(schemaName);
    return this.db.get<StoredSchema>(path);
  }

  /**
   * Check if a schema is registered
   *
   * @param schemaName The schema name
   * @returns True if the schema exists
   */
  async exists(schemaName: string): Promise<boolean> {
    const stored = await this.get(schemaName);
    return stored !== undefined;
  }

  /**
   * Update a schema (replaces existing)
   *
   * @param schema The updated schema
   */
  async update(schema: PraxisSchema): Promise<void> {
    await this.register(schema);
  }

  /**
   * List all registered schema names
   *
   * Implementation note: This method uses an index stored at `/_praxis/schemas/_index`.
   * When using InMemoryPraxisDB, schemas must be registered using `registerWithIndex()`
   * for them to appear in this listing. When using a full PluresDB implementation,
   * native listing capabilities should be used instead.
   *
   * @returns Array of registered schema names
   */
  async list(): Promise<string[]> {
    const indexPath = `${PRAXIS_PATHS.SCHEMAS}/_index`;
    const index = await this.db.get<string[]>(indexPath);
    return index ?? [];
  }

  /**
   * Register a schema and update the index
   *
   * @param schema The schema to register
   */
  async registerWithIndex(schema: PraxisSchema): Promise<void> {
    // Register the schema
    await this.register(schema);

    // Update the index
    const indexPath = `${PRAXIS_PATHS.SCHEMAS}/_index`;
    const existingIndex = (await this.db.get<string[]>(indexPath)) ?? [];

    if (!existingIndex.includes(schema.name)) {
      await this.db.set(indexPath, [...existingIndex, schema.name]);
    }
  }
}

/**
 * Register a schema in PluresDB
 *
 * Convenience function for one-off schema registration.
 *
 * @param db The PraxisDB instance
 * @param schema The schema to register
 *
 * @example
 * ```typescript
 * const db = createInMemoryDB();
 * await registerSchema(db, {
 *   version: "1.0.0",
 *   name: "MyApp",
 *   description: "My application schema"
 * });
 * ```
 */
export async function registerSchema(db: PraxisDB, schema: PraxisSchema): Promise<void> {
  const registry = new PraxisSchemaRegistry(db);
  await registry.registerWithIndex(schema);
}

/**
 * Create a PraxisSchemaRegistry instance
 *
 * @param db The PraxisDB instance
 * @returns PraxisSchemaRegistry instance
 */
export function createSchemaRegistry(db: PraxisDB): PraxisSchemaRegistry {
  return new PraxisSchemaRegistry(db);
}
